WordPress/Classes/Services/MediaService.m

```diff
@@ -485,17 +611,83 @@ - (void)getMediaLibraryServerCountForBlog:(Blog *)blog
+         */
+        Media *media = (Media *)[self.managedObjectContext objectWithID: mediaID];
+        if ([self supportsNewMediaExports]) {
+            [self.thumbnailService thumbnailURLForMedia:media
+                                          preferredSize:preferredSize
+                                           onCompletion:^(NSURL *url) {
+                                               completion(url, nil);
+                                           }
+                                                onError:^(NSError *error) {
+                                                    completion(nil, error);
+                                                }];
+        } else {
+            completion(media.absoluteThumbnailLocalURL, nil);
+        }
+    }];
 }
 
+- (void)thumbnailImageForMedia:(nonnull Media *)mediaInRandomContext
+                 preferredSize:(CGSize)preferredSize
+                    completion:(void (^)(UIImage * _Nullable image, NSError * _Nullable error))completion
+{
+    NSManagedObjectID *mediaID = [mediaInRandomContext objectID];
+    [self.managedObjectContext performBlock:^{
+        /*
+         When using the new Media export services, generate a thumbnail image from the URL
+         of the MediaThumbnailService.
+         Otherwise, use the legacy category method `imageForMedia:`.
+         */
+        Media *media = (Media *)[self.managedObjectContext objectWithID: mediaID];
+        if ([self supportsNewMediaExports]) {
+            [self.thumbnailService thumbnailURLForMedia:media
+                                          preferredSize:preferredSize
+                                           onCompletion:^(NSURL *url) {
+                                               dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0),^{
+                                                   UIImage *image = [UIImage imageWithContentsOfFile:url.path];
+                                                   [self.managedObjectContext performBlock:^{
+                                                       completion(image, nil);
+                                                   }];
+                                               });
+                                           }
+                                                onError:^(NSError *error) {
+                                                    completion(nil, error);
+                                                }];
+        } else {
+            [self imageForMedia:mediaInRandomContext
+                           size:preferredSize
+                        success:^(UIImage *image) {
+                            completion(image, nil);
+                        } failure:^(NSError *error) {
+                            completion(nil, error);
+                        }];
+        }
+    }];
+}
+
+- (MediaThumbnailService *)thumbnailService
+{
+    if (_thumbnailService) {
+        return _thumbnailService;
```

Comment:
Perhaps switch this over to check `if (!_thumbnailService)`, and create it if it doesn't exist – removes a `return`, that way round :)